libraries/chain_kv/include/b1/session/bytes.hpp

```diff
@@ -0,0 +1,284 @@
+      return result;
+    }
+
+    auto chunk_length = 3 * sizeof(size_t) + length;
+    auto* chunk = reinterpret_cast<char*>(a->malloc(chunk_length));
+    
+    result.m_memory_allocator_address = reinterpret_cast<size_t*>(chunk);
+    result.m_use_count_address = reinterpret_cast<size_t*>(chunk + sizeof(size_t));
+    result.m_length = reinterpret_cast<size_t*>(chunk + 2 * sizeof(size_t));
+    result.m_data = reinterpret_cast<size_t*>(chunk + 3 * sizeof(size_t));
+
+    *(result.m_memory_allocator_address) = reinterpret_cast<size_t>(&a->free_function());
+    *(result.m_use_count_address) = 1;
+    *(result.m_length) = length;
+    memcpy(result.m_data, data, length);
+
+    return result;
+}
+
+template <>
+inline bytes make_bytes(const void* data, size_t length) {
+    auto result = bytes{};
+    
+    result.m_memory_allocator_address = nullptr;
+    result.m_use_count_address = nullptr;
+    // encode the length as the pointer address of this data member.
+    result.m_length = reinterpret_cast<size_t*>(length);
+    result.m_data = reinterpret_cast<void*>(const_cast<void*>(data));
+    
+    return result;
+}
+
+// \brief Creates a new bytes instance with the given pointer and length.
+//
+// The returns a bytes "view" instance, in that the resulting bytes instance does not manage
+// the given memory.
+//
+// \tparam T The type stored in the array.
+// \param data A pointer to an array of items to store in a bytes instance.
+// \param length The number of items in the array.
+// \warning The resulting bytes instance does NOT take ownership of the given data pointer.
+// \warning The data pointer must remain in scope for the lifetime of the given bytes instance.
+template <typename T>
+bytes make_bytes(const T* data, size_t length) {
+    return make_bytes(reinterpret_cast<const void*>(data), length * sizeof(T));
+}
+
+inline const bytes bytes::invalid{};
+
+inline bytes::bytes(const bytes& b)
+: m_memory_allocator_address{b.m_memory_allocator_address},
+  m_use_count_address{b.m_use_count_address},
+  m_length{b.m_length},
+  m_data{b.m_data} {
+
+    if (m_use_count_address) {
+        ++(*m_use_count_address);
+    }
+}
+
+inline bytes::bytes(bytes&& b)
+: m_memory_allocator_address{std::move(b.m_memory_allocator_address)},
+  m_use_count_address{std::move(b.m_use_count_address)},
+  m_length{std::move(b.m_length)},
+  m_data{std::move(b.m_data)} {
+
+    b.m_memory_allocator_address = nullptr;
+    b.m_use_count_address = nullptr;
+    b.m_length = nullptr;
+    b.m_data = nullptr;
+}
+
+inline bytes::~bytes() {
+    if (!m_data || !m_length || !m_use_count_address) {
+        return;
+    }
+    
+    if (--(*m_use_count_address) != 0) {
+        return;
+    }
+    
+    // The memory pool must remain in scope for the lifetime of this instance.
+    auto* free_function = reinterpret_cast<free_function_type*>(*m_memory_allocator_address);
+    (*free_function)(m_data, 3 * sizeof(size_t) + *m_length);
```

Comment:
Why do you need this indirection?  Couldn't m_memory_allocator_address be a `free_function_type*`?